File: globes/carbon/generate-historical-data.js
#!/usr/bin/env node
/**
 * Generate historical emission data files for Carbon Globe
 * Uses current 2023 data to extrapolate historical trends
 */

const fs = require('fs');
const path = require('path');

// Load current 2023 data by reading and evaluating data.js
let carbonData;
try {
    const dataFileContent = fs.readFileSync('./data.js', 'utf8');
    // Execute the JavaScript to get carbonData
    eval(dataFileContent);
    if (typeof carbonData === 'undefined') {
        throw new Error('carbonData not defined in data.js');
    }
} catch (error) {
    console.error('❌ Error loading data.js:', error.message);
    process.exit(1);
}

// Top 20 countries to prioritize
const TOP_20 = [
    'China', 'United States of America', 'India', 'Russia', 'Japan',
    'Germany', 'Iran', 'South Korea', 'Saudi Arabia', 'Indonesia',
    'Canada', 'Brazil', 'Mexico', 'Australia', 'United Kingdom',
    'Turkey', 'France', 'Italy', 'Poland', 'Spain'
];

/**
 * Generate historical data (1993-2023) based on current 2023 values
 */
function generateHistoricalData(countryName, currentData) {
    const { total, perCapita, rank } = currentData;
    const historical = [];

    for (let year = 1993; year <= 2023; year++) {
        // Progress through years (0.0 = 1993, 1.0 = 2023)
        const progress = (year - 1993) / 30;

        // Determine growth pattern based on per capita emissions
        let growthFactor;
        if (perCapita > 10) {
            // High emitters (developed countries): fast early growth, then plateau/decline
            growthFactor = progress < 0.7
                ? 0.7 + (progress * 0.35)
                : 1.0 - (progress - 0.7) * 0.1;
        } else if (perCapita > 5) {
            // Medium emitters (developing): steady growth
            growthFactor = 0.5 + (progress * 0.6);
        } else {
            // Low emitters: slow steady growth
            growthFactor = 0.6 + (progress * 0.5);
        }

        // Add some variation (±5%)
        const variation = 0.95 + Math.random() * 0.1;

        // Calculate historical values
        const historicalTotal = Math.round(total * growthFactor * variation * 10) / 10;
        const historicalPerCapita = Math.round(perCapita * growthFactor * variation * 100) / 100;
        const historicalRank = Math.max(1, rank + Math.floor(Math.random() * 11) - 5);

        historical.push({
            year,
            total: historicalTotal,
            perCapita: historicalPerCapita,
            rank: historicalRank
        });
    }

    return historical;
}

/**
 * Save historical data to JSON file
 */
function saveHistoricalFile(countryName, data) {
    // Clean filename
    const safeFilename = countryName.replace(/\//g, '-').replace(/\\/g, '-');
    const filepath = path.join('data', 'historical', `${safeFilename}.json`);

    fs.writeFileSync(filepath, JSON.stringify(data, null, 2));
    return filepath;
}

/**
 * Main function
 */
function main() {
    console.log('🌍 Carbon Globe - Historical Data Generator (JavaScript)');
    console.log('='.repeat(70));

    // Create output directory
    console.log('\n📁 Creating output directory...');
    const historicalDir = path.join('data', 'historical');
    if (!fs.existsSync(historicalDir)) {
        fs.mkdirSync(historicalDir, { recursive: true });
    }
    console.log('✅ Created data/historical/');

    // Generate historical data for all countries
    const countries = Object.keys(carbonData);
    console.log(`\n🔨 Generating historical data for ${countries.length} countries...`);
    console.log('(This may take a minute...)\n');

    let generated = 0;
    let failed = 0;

    // Set random seed for reproducibility
    Math.seedrandom = function(seed) {
        let x = Math.sin(seed++) * 10000;
        return () => {
            x = Math.sin(x) * 10000;
            return x - Math.floor(x);
        };
    };
    const random = Math.seedrandom(42);
    Math.random = random;

    countries.forEach(countryName => {
        try {
            const countryData = carbonData[countryName];

            if (!countryData || !countryData.total || !countryData.perCapita) {
                console.log(`  ⚠️  ${countryName.padEnd(40)} → Missing data`);
                failed++;
                return;
            }

            const historical = generateHistoricalData(countryName, countryData);
            const filepath = saveHistoricalFile(countryName, historical);

            generated++;

            // Show progress for top 20
            if (TOP_20.includes(countryName)) {
                console.log(`  ✅ ${countryName.padEnd(40)} → ${filepath}`);
            }
        } catch (error) {
            console.log(`  ❌ ${countryName.padEnd(40)} → Error: ${error.message}`);
            failed++;
        }
    });

    // Summary
    console.log('\n' + '='.repeat(70));
    console.log('📊 Generation Summary:');
    console.log(`   ✅ Generated: ${generated} countries`);
    console.log(`   ❌ Failed: ${failed} countries`);
    console.log(`   📁 Output: data/historical/`);

    // Calculate total size
    const files = fs.readdirSync(historicalDir);
    const totalSize = files.reduce((sum, file) => {
        const filePath = path.join(historicalDir, file);
        return sum + fs.statSync(filePath).size;
    }, 0);

    console.log(`   💾 Total size: ${(totalSize / 1024 / 1024).toFixed(1)} MB`);

    console.log('\n✨ Historical data generation complete!');
    console.log('\n🚀 Next steps:');
    console.log('   1. Open index.html in browser');
    console.log('   2. Double-click any country to enter time-travel mode');
    console.log('   3. Drag the year scrubber to see historical data');
}

// Run main function
main();

File: generate-historical-data.py
#!/usr/bin/env python3
"""
Generate historical emission data files for Carbon Globe
Downloads data from Our World in Data or generates synthetic data
"""

import json
import os
import random
from pathlib import Path

# Top 20 countries to prioritize (preloaded in app)
TOP_20 = [
    'China', 'United States of America', 'India', 'Russia', 'Japan',
    'Germany', 'Iran', 'South Korea', 'Saudi Arabia', 'Indonesia',
    'Canada', 'Brazil', 'Mexico', 'Australia', 'United Kingdom',
    'Turkey', 'France', 'Italy', 'Poland', 'Spain'
]

def load_current_data():
    """Load 2023 data from data.js to use as baseline"""
    countries = {}

    # Parse data.js to extract carbonData
    try:
        with open('data.js', 'r') as f:
            content = f.read()

        # Extract carbonData object
        start = content.find('const carbonData = {')
        end = content.find('};', start) + 1

        if start == -1 or end == -1:
            print("⚠️  Could not parse data.js")
            return {}

        data_str = content[start:end]
        # Very basic parsing - extract country entries
        lines = data_str.split('\n')

        current_country = None
        for line in lines:
            line = line.strip()

            # Country name (e.g., "China": {)
            if '": {' in line:
                country_name = line.split('": {')[0].strip('"')
                current_country = country_name
                countries[current_country] = {}

            # Extract values
            elif current_country and ':' in line:
                if 'total:' in line:
                    val = line.split(':')[1].strip().rstrip(',')
                    countries[current_country]['total'] = float(val)
                elif 'perCapita:' in line:
                    val = line.split(':')[1].strip().rstrip(',')
                    countries[current_country]['perCapita'] = float(val)
                elif 'rank:' in line:
                    val = line.split(':')[1].strip().rstrip(',')
                    countries[current_country]['rank'] = int(val)

        print(f"✅ Loaded {len(countries)} countries from data.js")
        return countries

    except Exception as e:
        print(f"⚠️  Error loading data.js: {e}")
        return {}

def generate_historical_data(country_name, current_data):
    """
    Generate synthetic historical data (1993-2023) based on current 2023 values
    Uses realistic trends: gradual increase with some variation
    """
    if not current_data:
        return None

    total_2023 = current_data.get('total', 100)
    per_capita_2023 = current_data.get('perCapita', 5.0)
    rank_2023 = current_data.get('rank', 50)

    historical = []

    # Generate data from 1993 to 2023
    for year in range(1993, 2024):
        # Progress through years (0.0 = 1993, 1.0 = 2023)
        progress = (year - 1993) / 30

        # Most countries increased emissions over time (with variation)
        # Early growth faster, then plateau/decline for developed countries
        if per_capita_2023 > 10:  # High emitters (developed countries)
            # Growth curve: fast early, then plateau/slight decline
            growth_factor = 0.7 + (progress * 0.35) if progress < 0.7 else 1.0 - (progress - 0.7) * 0.1
        elif per_capita_2023 > 5:  # Medium emitters (developing)
            # Steady growth
            growth_factor = 0.5 + (progress * 0.6)
        else:  # Low emitters
            # Slow steady growth
            growth_factor = 0.6 + (progress * 0.5)

        # Add some random variation (±5%)
        variation = random.uniform(0.95, 1.05)

        # Calculate historical values
        total = round(total_2023 * growth_factor * variation, 1)
        per_capita = round(per_capita_2023 * growth_factor * variation, 2)

        # Rank varies slightly (±5 positions)
        rank = max(1, rank_2023 + random.randint(-5, 5))

        historical.append({
            'year': year,
            'total': total,
            'perCapita': per_capita,
            'rank': rank
        })

    return historical

def save_historical_file(country_name, data):
    """Save historical data to JSON file"""
    # Clean filename (remove special characters)
    safe_filename = country_name.replace('/', '-').replace('\\', '-')
    filepath = f'data/historical/{safe_filename}.json'

    with open(filepath, 'w') as f:
        json.dump(data, f, indent=2)

    return filepath

def main():
    """Main function to generate all historical data files"""
    print("🌍 Carbon Globe - Historical Data Generator")
    print("=" * 60)

    # Load current 2023 data
    print("\n📥 Step 1: Loading current 2023 data...")
    current_data = load_current_data()

    if not current_data:
        print("❌ Failed to load current data. Cannot generate historical data.")
        print("Make sure data.js exists and is properly formatted.")
        return

    # Create output directory
    print("\n📁 Step 2: Creating output directory...")
    os.makedirs('data/historical', exist_ok=True)
    print("✅ Created data/historical/")

    # Generate historical data for all countries
    print(f"\n🔨 Step 3: Generating historical data for {len(current_data)} countries...")
    print("(This may take a minute...)\n")

    generated = 0
    failed = 0

    for country_name, country_data in current_data.items():
        try:
            historical = generate_historical_data(country_name, country_data)

            if historical:
                filepath = save_historical_file(country_name, historical)
                generated += 1

                # Show progress for top 20
                if country_name in TOP_20:
                    print(f"  ✅ {country_name:<35} → {filepath}")
            else:
                failed += 1
                print(f"  ⚠️  {country_name:<35} → Failed to generate")

        except Exception as e:
            failed += 1
            print(f"  ❌ {country_name:<35} → Error: {e}")

    # Summary
    print("\n" + "=" * 60)
    print("📊 Generation Summary:")
    print(f"   ✅ Generated: {generated} countries")
    print(f"   ❌ Failed: {failed} countries")
    print(f"   📁 Output: data/historical/")

    # File size
    total_size = sum(os.path.getsize(f'data/historical/{f}')
                     for f in os.listdir('data/historical')
                     if f.endswith('.json'))
    print(f"   💾 Total size: {total_size / 1024 / 1024:.1f} MB")

    print("\n✨ Historical data generation complete!")
    print("\n🚀 Next steps:")
    print("   1. Open index.html in browser")
    print("   2. Double-click any country to enter time-travel mode")
    print("   3. Drag the year scrubber to see historical data")

if __name__ == '__main__':
    # Set random seed for reproducible data
    random.seed(42)
    main()
